src/openlrr/game/world/Roof.cpp

```diff
@@ -22,30 +24,304 @@ LegoRR::Roof_Globs & LegoRR::roofGlobs = *(LegoRR::Roof_Globs*)0x00553180;
+								vertices[i].normal.z = InvTemp * vertices[i].normal.z;
+								vertices[i].normal.x = InvTemp * vertices[i].normal.x;
+								vertices[i].normal.y = InvTemp * vertices[i].normal.y;
+							}
+							Gods98::Container_Mesh_SetVertices(roofGlobs.contMesh, locRoofBlock->groupID, 0, 4, vertices);
+							/* &= ~ROOF_NEEDUPDATE; // (~0x2) updated, so clear flag */
+							locRoofBlock->flags &= ~ROOF_FLAG_NEEDUPDATE;
+							width = roofGlobs.dimensions.width;
+						}
+					}
+				}
+			}
+		}
+	}
+}
 
 // <LegoRR.exe @00463770>
-//void __cdecl LegoRR::Roof_Hide(bool32 hide);
+void __cdecl LegoRR::Roof_Hide(bool32 hide)
+{
+	// Check if roof is already hidden, if not, hide it
+	if (roofGlobs.hidden != hide) {
+		roofGlobs.hidden = hide;
+		if (hide != 0) {
+			Gods98::Container_Hide(roofGlobs.contMesh, TRUE);
+			return;
+		}
+		Container_Hide(roofGlobs.contMesh, FALSE);
+	}
+}
 
 // <LegoRR.exe @004637b0>
-//void __cdecl LegoRR::Roof_AddVisibleBlock(uint32 bx, uint32 by);
+void __cdecl LegoRR::Roof_AddVisibleBlock(uint32 bx, uint32 by)
+{
+	int idx;
+	
+	idx = (by * roofGlobs.dimensions.width) + bx;
+	Gods98::Container_Mesh_HideGroup(roofGlobs.contMesh, roofGlobs.grid[idx].groupID, FALSE);
+	roofGlobs.visibleTable[roofGlobs.visibleCount] = roofGlobs.grid + idx;
+	roofGlobs.visibleCount++;
+}
 
 // <LegoRR.exe @00463800>
-//void __cdecl LegoRR::Roof_HideAllVisibleBlocks(void);
+void __cdecl LegoRR::Roof_HideAllVisibleBlocks(void)
+{
+	// If any blocks visable clear them out
+	if (roofGlobs.visibleCount != 0) {
+		for(uint32 i = 0; i < roofGlobs.visibleCount; i++) {
+			Gods98::Container_Mesh_HideGroup(roofGlobs.contMesh, roofGlobs.visibleTable[i]->groupID, TRUE);
+		}
+	}
+	roofGlobs.visibleCount = 0;
+}
 
 // <LegoRR.exe @00463850>
-//bool32 __cdecl LegoRR::Roof_GetBlockPlaneNormal(uint32 bx, uint32 by, OUT Vector3F* normal);
+bool32 __cdecl LegoRR::Roof_GetBlockPlaneNormal(uint32 bx, uint32 by, OUT Vector3F* normal)
+{
+	float fVar1;
+	float fVar2;
+	float fVar3;
+	float fVar4;
+	Vector3F normal1;
+	Vector3F normal2;
+	Vertex vertices[4];
+
+	// Check if block is within bounds, calc normals, then assign to caller object
+	if (bx < (uint32)roofGlobs.dimensions.width && by < (uint32)roofGlobs.dimensions.height && (roofGlobs.grid[by * roofGlobs.dimensions.width + bx].flags & ROOF_FLAG_HIDDEN) == 0) {
+		Gods98::Container_Mesh_GetVertices(roofGlobs.contMesh,roofGlobs.grid[by * roofGlobs.dimensions.width + bx].groupID, 0, 4, vertices);
+		Gods98::Maths_PlaneNormal(&normal1, &vertices[0].position, &vertices[1].position, &vertices[2].position);
+		Gods98::Maths_PlaneNormal(&normal2, &vertices[2].position, &vertices[3].position, &vertices[0].position);
+		fVar3 = normal2.y + normal1.y;
+		fVar1 = normal2.z + normal1.z;
+		fVar2 = normal2.x + normal1.x;
+		fVar4 = 1.0 / sqrtf(fVar2 * fVar2 + fVar1 * fVar1 + fVar3 * fVar3);
+		normal->x = fVar4 * fVar2;
+		normal->y = fVar4 * fVar3;
+		normal->z = fVar4 * fVar1;
```

Comment:
These 7 lines can be simplified to. It's not necessary to always recreate the Maths logic with function calls, but I'll suggest it if there are other changes that need to be made anyway.
```cpp
Vector3F normalAvg;
Gods98::Maths_Vector3DAdd(&normalAvg, &normal1, &normal2);
Gods98::Maths_Vector3DNormalize(&normalAvg);
*normal = normalAvg;
```
You can also substitute `normalAvg` with re-using `normal1` or `normal2`. There won't be any side effects because PlaneNormal will always assign to the result value.